### PR TITLE
Add metrics

### PR DESCRIPTION
**Pull request by gpaul**
This PR adds metrics reporting over statsd/UDP.

TODO:

- [x] Add initial metrics framework
- [x] Add statsd reporting over UDP (gated on `-statsd-udp-host-env-var`) being set by the user
- [x] Report `uptime`
- [x] Report duration of RPCs (use gRPC unary interceptor if possible)
- [x] Report RPC successes
- [x] Report RPC failures
- [x] Update README

Future work:
- [ ] Add support for prometheus (gated on `-prometheus-listen-addr` being set by the user)

Fixes https://jira.mesosphere.com/browse/DCOS_OSS-4504

### CODE REVIEW

**Comment by jdef** [comment]
can we track prometheus support in a separate ticket (and tackle it later)?

**Comment by gpaul** [comment]
Tested manually in soak. Works like a charm.

File: cmd/csilvm/csilvm.go [hunk 1]
```diff
@@ -84,13 +95,66 @@ func main() {
...
+	scope := tally.NoopScope
+	if *statsdUDPHostEnvVarF != "" && *statsdUDPPortEnvVarF != "" {
+		statsdHost := os.Getenv(*statsdUDPHostEnvVarF)
+		statsdPort := os.Getenv(*statsdUDPPortEnvVarF)
+		statsdServerAddr := fmt.Sprintf("%s:%s", statsdHost, statsdPort)
+		// Set no statsd prefix, tags are already prefixed using 'csilvm'.
+		const (
+			statsdPrefix     = ""
+			maxFlushInterval = time.Second
+			maxUDPPacketSize = 1440
```
Comment (jdef) [inline cmd/csilvm/csilvm.go]: should this be configurable?

File: pkg/csilvm/metrics_test.go [hunk 2]
```diff
@@ -0,0 +1,327 @@
...
+			return timerSnapshot, true
+		}
+	}
+	return nil, false
+}
+
+func (m timerMap) mustGet(name string, opts ...getOpt) tally.TimerSnapshot {
+	timer, ok := m.get(name, opts...)
+	if !ok {
+		panic(fmt.Sprintf("cannot find timer %q", name))
```
Comment (jdef) [inline pkg/csilvm/metrics_test.go]: if `mustGet` isn't ever invoked in a parallel goroutine, prefer `t.Fatal` to `panic`

File: pkg/csilvm/metrics_test.go [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+			return counterSnapshot, true
+		}
+	}
+	return nil, false
+}
+
+func (m counterMap) mustGet(name string, opts ...getOpt) tally.CounterSnapshot {
+	counter, ok := m.get(name, opts...)
+	if !ok {
+		panic(fmt.Sprintf("cannot find counter %q", name))
```
Comment (jdef) [inline pkg/csilvm/metrics_test.go]: ditto re: t.Fatal vs. panic

File: pkg/csilvm/metrics_test.go [hunk 4]
```diff
@@ -0,0 +1,327 @@
...
+			return gaugeSnapshot, true
+		}
+	}
+	return nil, false
+}
+
+func (m gaugeMap) mustGet(name string, opts ...getOpt) tally.GaugeSnapshot {
+	gauge, ok := m.get(name, opts...)
+	if !ok {
+		panic(fmt.Sprintf("cannot find gauge %q", name))
```
Comment (jdef) [inline pkg/csilvm/metrics_test.go]: ditto

File: pkg/csilvm/metrics.go [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+	"google.golang.org/grpc"
+)
+
+func MetricsInterceptor(scope tally.Scope) grpc.UnaryServerInterceptor {
+	scope = scope.SubScope("requests")
+	return func(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
+		scope = scope.Tagged(map[string]string{
+			"method": info.FullMethod,
+		})
+		timer := scope.Timer("duration")
```
Comment (jdef) [inline pkg/csilvm/metrics.go]: looks like tally has something nice and reusable https://github.com/uber-go/tally/blob/master/instrument/call.go

File: cmd/csilvm/csilvm.go (hunk 1 above)
Comment (gpaul) [reply cmd/csilvm/csilvm.go]: Yeah - I think so - especially if the metrics server is not local ... in which case we will want to set this to 512. Also - I see the library prefers a slightly smaller value than 1440.

File: pkg/csilvm/metrics_test.go (hunk 2 above)
Comment (gpaul) [reply pkg/csilvm/metrics_test.go]: I didn't want to push `*testing.T` into these types, but I guess there's no getting away from it.

Comment (gpaul) [reply pkg/csilvm/metrics_test.go]: `*testing.T` is the `context` of the `_test.go` world.

File: pkg/csilvm/metrics.go (hunk 5 above)
Comment (gpaul) [reply pkg/csilvm/metrics.go]: Nice, i'll use that

Comment (gpaul) [reply pkg/csilvm/metrics.go]: Hmm - that results in some awkward code and metrics called "/csi..." etc. I think I'll switch `duration` to `latency` and use `result_type: success/error`, but not switch to using their `instrument.Call` abstraction.

**Comment by gpaul** [comment]
I'll have to test this on a cluster next week.

**Pull request merged**